### PR TITLE
OK action is enabled if all members are deselected #168

### LINKED ISSUE #168: OK action is enabled if all members are deselected
**Description**

OK action should be disabled in plugin UI if all members are deselected in members table.
When dialog is opened, something is always checked: all methods or one of them. But user can deselect all.

### PR DESCRIPTION
**Pull request by Vassiliy-Kudryashov**
# Description

Now OK button is disabled if no members is selected (ticked), better validation and hints were added for:
1. Not configured test root directory
2. Not valid test root directory (for example a directory out of project scope)
3. Empty members (methods) selection

Fixes #168 

## Type of Change
- Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

## Manual Scenario 

1. Call "Generate action" on some class with methods. 
4. Delselect all methods (Ctrl+A, then Space)
5. Observe disabled OK button.

# Checklist (remove irrelevant options):

- [x] The change followed the style guidelines of the UTBot project
- [x] Self-review of the code is passed
- [x] The change contains enough commentaries, particularly in hard-to-understand areas
- [x] No new warnings

### CODE REVIEW

File: utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt [hunk 1]
```diff
@@ -418,20 +436,18 @@ class GenerateTestsDialogWindow(val model: GenerateTestsModel) : DialogWrapper(m
...
-        val rootExists = getOrCreateTestRoot(testSourceRoot)
-        if (rootExists) {
-            if (cbSpecifyTestPackage.isSelected) {
-                createSelectedPackage(testSourceRoot)
-            } else {
-                createPackagesByClasses(testSourceRoot)
-            }
-            return true
+        if (cbSpecifyTestPackage.isSelected) {
+            createSelectedPackage(sourceRootAndEntry.first)
```
[the remark this answers is not in the document]
Comment (EgorkaKulikov) [reply utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt:446]: It seems that you do not need second here at all...

File: utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt [hunk 2]
```diff
@@ -460,19 +476,14 @@ class GenerateTestsDialogWindow(val model: GenerateTestsModel) : DialogWrapper(m
...
-            .firstOrNull { VfsUtil.isAncestor(it.file!!, testSourceRoot, true) }
-            ?: return false
-
-        VfsUtil.createDirectoryIfMissing(urlToPath(testSourceRoot.url))
-
-        contentEntry.addSourceFolder(testSourceRoot.url, codegenLanguages.item.testRootType())
-        WriteCommandAction.runWriteCommandAction(model.project) { modifiableModel.commit() }
+    private fun getRootDirectoryAndContentEntry() : Pair<VirtualFile, ContentEntry>? {
+        val testSourceRoot = model.testSourceRoot ?: return null
+        if (!testSourceRoot.isDirectory) return null
```
Comment (EgorkaKulikov) [inline utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt:481]: Can't we incapsulate this logic just in one place, for example, in doValidate method?

File: utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt [hunk 3]
```diff
@@ -345,6 +341,28 @@ class GenerateTestsDialogWindow(val model: GenerateTestsModel) : DialogWrapper(m
 
     private fun checkMembers(members: List<MemberInfo>) = members.forEach { it.isChecked = true }
 
+    override fun doValidate(): ValidationInfo? {
+        val testSourceRoot = model.testSourceRoot
+        if (testSourceRoot == null || !testSourceRoot.isDirectory) {
+            return ValidationInfo("Test source root is not configured", testSourceFolderField.childComponent)
+        }
+        if (getRootDirectoryAndContentEntry() == null) {
```
Comment (EgorkaKulikov) [inline utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt:349]: Actually here you check just it's "second" as "first" is checked in the previous condition.

File: utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt [hunk 4]
```diff
@@ -345,6 +341,28 @@ class GenerateTestsDialogWindow(val model: GenerateTestsModel) : DialogWrapper(m
 
     private fun checkMembers(members: List<MemberInfo>) = members.forEach { it.isChecked = true }
 
+    override fun doValidate(): ValidationInfo? {
+        val testSourceRoot = model.testSourceRoot
+        if (testSourceRoot == null || !testSourceRoot.isDirectory) {
+            return ValidationInfo("Test source root is not configured", testSourceFolderField.childComponent)
```
Comment (EgorkaKulikov) [inline utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt:347]: Actually we have another validator for it.

**Review by EgorkaKulikov — CHANGES_REQUESTED** [review]

File: utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (EgorkaKulikov) [reply utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt:446]: val (a, b) = getAOrb()

File: utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt [hunk 5]
```diff
@@ -460,19 +476,14 @@ class GenerateTestsDialogWindow(val model: GenerateTestsModel) : DialogWrapper(m
...
-        contentEntry.addSourceFolder(testSourceRoot.url, codegenLanguages.item.testRootType())
-        WriteCommandAction.runWriteCommandAction(model.project) { modifiableModel.commit() }
+    private fun getRootDirectoryAndContentEntry() : Pair<VirtualFile, ContentEntry>? {
+        val testSourceRoot = model.testSourceRoot ?: return null
+        if (!testSourceRoot.isDirectory) return null
 
-        return true
+        val contentEntry = ModuleRootManager.getInstance(model.testModule).contentEntries
+            .filterNot { it.file == null }
+            .firstOrNull { VfsUtil.isAncestor(it.file!!, testSourceRoot, true) } ?: return null
```
[the remark this answers is not in the document]
Comment (Vassiliy-Kudryashov) [reply utbot-intellij/src/main/kotlin/org/utbot/intellij/plugin/ui/GenerateTestsDialogWindow.kt:485]: We shouldn't assign null to contentEntry, null means method have to return null

**Review by EgorkaKulikov — APPROVED** [review]

**Pull request merged**